### PR TITLE
Fixing some issues with docstrings and adding CI/CD step

### PR DESCRIPTION
**Pull request by devendragovil**
This PR does two things:

1. It fixes some docstrings issues with various modules in the library. 
2. It adds an explicit CI/CD step where now numpydoc will validate docstrings to make sure that they are in compliance, otherwise the CI/CD should fail. 

The second is in my opinion important because otherwise the docstrings are not compliant to numpy, and in many cases is not reliably parsed by Sphinx while generating docs. 

The Sphinx Build step does do a similar check, however it only throws warnings and not errors. This additional step can be removed once we integrate numpydoc validation directly with Sphinx by enabling it to throw exceptions instead of warnings. 

This helps with #178

### CODE REVIEW

**Comment by devendragovil** [comment]
I have added the numpydoc-validation as an additional job in the docs workflow to help it isolate from the rest. 

It won't pass for now, as there are a lot of issues in the docstrings that need to be resolved. 

One can get a list of all issues by doing the following:

1. Clone/Pull the repo to sync with upstream.
2. install the new pre-commits.
3. Run the following command: `pre-commit run --all-files numpydoc-validation`
4. To store in a file you can run `pre-commit run --all-files numpydoc-validation &> all_warnings_numpydoc.txt`

I would need everyone's help to resolve #178

**Review by ffl096 — CHANGES_REQUESTED** [review]
Having looked at this more, I now think using numpydoc validation in a separate step instead of inside Sphinx is a better option indeed. Sphinx has some notorious caching behaviour that swallows warnings on my machine and hinders local validation.

File: toponetx/algorithms/components.py [hunk 1]
```diff
@@ -174,15 +174,16 @@ def connected_components(
...
 
-    Yields
-    ------
-    set[Hashable] or set[tuple[Hashable, ...]]
-        Returns subcomplexes generated by the cells (or nodes) in the
+    Returns
+    -------
+    Generator[set[Hashable] | set[tuple[Hashable, ...]], None, None]
+        Returns a generator of subcomplexes generated by the cells (or nodes) in the
         cell(node) components of complex.
```
[the remark this answers is not in the document]
Comment (devendragovil) [reply toponetx/algorithms/components.py:181]: Yields is a correction section however numpydoc requires a Returns section, and raises an error without that. I am implementing both the Returns and Yields section for now.

File: .github/workflows/docs.yml [hunk 2]
```diff
@@ -43,4 +43,31 @@ jobs:
...
+          python-version: ${{matrix.python-version}}
+          cache: "pip"
+          cache-dependency-path: "pyproject.toml"
+      - name: Install dependencies [pip]
+        run: |
+          pip install --upgrade pip setuptools wheel
+          pip install -e .[doc]
+      - name: Install Pandoc [apt-get]
+        run: |
+          sudo apt-get -y install pandoc
```
[the remark this answers is not in the document]
Comment (devendragovil) [reply .github/workflows/docs.yml:68]: Fixed in new commit.

File: toponetx/algorithms/spectrum.py [hunk 3]
```diff
@@ -65,7 +65,7 @@ def hodge_laplacian_eigenvectors(
 
     Returns
     -------
-    k_eigenvals, k_eigenvectors : numpy.ndarray
+    numpy.ndarray
         First k eigevals and eigenvec associated with the hodge laplacian matrix.
```
[the remark this answers is not in the document]
Comment (devendragovil) [reply toponetx/algorithms/spectrum.py:69]: Fixed in new commit.

File: toponetx/algorithms/spectrum.py [hunk 4]
```diff
@@ -161,7 +161,7 @@ def laplacian_beltrami_eigenvectors(
 
     Returns
     -------
-    eigenvectors, eigenvalues : numpy.ndarray
+    numpy.ndarray
```
[the remark this answers is not in the document]
Comment (devendragovil) [reply toponetx/algorithms/spectrum.py:164]: Fixed in new commit.

File: toponetx/algorithms/components.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (devendragovil) [reply toponetx/algorithms/components.py:181]: It doesn't work unfortunately, it raises an error saying that there's both a Returns and Yields section. With only the Yields section, it raises an error saying that there is no Return section found. 

I think I have found the issue, this particular method has no yield keyword, as it is simply returning a generator from a different keyword. I think Returns hence is perhaps more accurate here.

[the remark this answers is not in the document]
Comment (ffl096) [reply toponetx/algorithms/components.py:181]: I propose to change the method body to
```python3
yield from s_connected_components(domain, s=1, cells=cells, return_singletons=True)
```

This should make validation happy and we can keep the `Yields` section.

[the remark this answers is not in the document]
Comment (devendragovil) [reply toponetx/algorithms/components.py]: Done!

File: toponetx/algorithms/components.py [hunk 5]
```diff
@@ -176,13 +176,14 @@ def connected_components(
 
     Yields
     ------
-    set[Hashable] or set[tuple[Hashable, ...]]
-        Returns subcomplexes generated by the cells (or nodes) in the
+    Generator[set[Hashable] | set[tuple[Hashable, ...]], None, None]
```
[the remark this answers is not in the document]
Comment (devendragovil) [reply toponetx/algorithms/components.py]: Done!

**Review by ffl096 — APPROVED** [review]

**Pull request merged**